utils/src/test/java/uk/gov/pay/commons/utils/logging/LoggingFilterTest.java

```diff
@@ -123,6 +125,40 @@ public void shouldLogEntryAndExitPointsEvenWhenFilterChainingThrowsException() t
         assertThat(endLogMessage, matchesRegex(".*total time \\d+ms"));
     }
 
+    @Test
+    public void shouldSetDiagnosticContextPerRequest() {
+        String requestUrl = "/publicauth-request";
+        String requestMethod = "GET";
+
+        when(mockRequest.getRequestURI()).thenReturn(requestUrl);
+        when(mockRequest.getMethod()).thenReturn(requestMethod);
+
+        when(mockRequest.getHeader("X-Request-Id")).thenReturn("some-id");
+        loggingFilter.doFilter(mockRequest, mockResponse, mockFilterChain);
+        assertThat(MDC.get("X-Request-Id"), is("some-id"));
+
+        when(mockRequest.getHeader("X-Request-Id")).thenReturn("some-other-id");
+        loggingFilter.doFilter(mockRequest, mockResponse, mockFilterChain);
+        assertThat(MDC.get("X-Request-Id"), is("some-other-id"));
+    }
+
+    @Test
+    public void shouldNotAddRequestIdToMDCIfMissing() {
+        String requestUrl = "/publicauth-request";
+        String requestMethod = "GET";
+
+        when(mockRequest.getRequestURI()).thenReturn(requestUrl);
+        when(mockRequest.getMethod()).thenReturn(requestMethod);
+
+        when(mockRequest.getHeader("X-Request-Id")).thenReturn("some-id");
+        loggingFilter.doFilter(mockRequest, mockResponse, mockFilterChain);
+        assertThat(MDC.get("X-Request-Id"), is("some-id"));
+
+        when(mockRequest.getHeader("X-Request-Id")).thenReturn(null);
+        loggingFilter.doFilter(mockRequest, mockResponse, mockFilterChain);
+        assertNull(MDC.get("X-Request-Id"));
+    }
+
```

Comment:
Not too sure about how these tests are separated. The first block of stubbing and assertions is identical in both of them.

I think what we’re really getting at is that `MDC` handles each request having its own request ID. I think this could be tested in a single test:

```java
    @Test
    public void shouldAddRequestIdsToMdc() {
        when(mockRequest.getRequestURI()).thenReturn("/publicauth-request");
        when(mockRequest.getMethod()).thenReturn(GET);

        when(mockRequest.getHeader("X-Request-Id")).thenReturn("some-id");
        loggingFilter.doFilter(mockRequest, mockResponse, mockFilterChain);
        assertThat(MDC.get("X-Request-Id"), is("some-id"));

        when(mockRequest.getHeader("X-Request-Id")).thenReturn("some-other-id");
        loggingFilter.doFilter(mockRequest, mockResponse, mockFilterChain);
        assertThat(MDC.get("X-Request-Id"), is("some-other-id"));

        when(mockRequest.getHeader("X-Request-Id")).thenReturn(null);
        loggingFilter.doFilter(mockRequest, mockResponse, mockFilterChain);
        assertThat(MDC.get("X-Request-Id"), is(null));
}
```